### PR TITLE
synapse_workspace_resource support azure_devops_repo, and github_repo

### LINKED ISSUE #10091: Support for Azure Synapse Workspace Git Configuration
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

The GA release of Azure Synapse Analytics workspaces supports Git Configurations similar to Data Factory.  This is a critical feature for most big data architecture platforms.

### New or Affected Resource(s)

* azurerm_synapse_workspace

### Potential Terraform Configuration

```hcl
resource "azurerm_synapse_workspace" "example" {
  name                                 = "example"
  resource_group_name                  = azurerm_resource_group.example.name
  location                             = azurerm_resource_group.example.location
  storage_data_lake_gen2_filesystem_id = azurerm_storage_data_lake_gen2_filesystem.example.id
  sql_administrator_login              = "sqladminuser"
  sql_administrator_login_password     = "H@Sh1CoR3!"

  aad_admin {
    login     = "AzureAD Admin"
    object_id = "00000000-0000-0000-0000-000000000000"
    tenant_id = "00000000-0000-0000-0000-000000000000"
  }

  **vsts_configuration {
      account_name    = "org_name"
      branch_name     = "master"
      project_name    = "project_name"
      repository_name = "repo_name"
      root_folder     = "/"
      tenant_id       = "00000000-0000-0000-0000-000000000000"
  }

   github_configuration {
      account_name    = "org_name"
      branch_name     = "master"
      repository_name = "repo_name"
      root_folder     = "/"
      git_url      = "https://github.mydomain.com"
  }**

  tags = {
    Env = "production"
  }
}
```

### References

* https://docs.microsoft.com/en-us/rest/api/synapse/workspaces/createorupdate#workspacerepositoryconfiguration
* https://docs.microsoft.com/en-us/azure/synapse-analytics/cicd/source-control#:~:text=Azure%20Synapse%20Studio%20git%20integration%20is%20not%20available,with%20only%20one%20git%20repository%20at%20a%20time.

### PR DESCRIPTION
**Pull request by allantargino**
fix #10091

It follows the same interface for source control from [`azurerm_data_factory`](https://github.com/terraform-providers/terraform-provider-azurerm/blob/62b6a2e0d5b9906fd838bbab9f5f7da6a5e18421/azurerm/internal/services/datafactory/data_factory_resource.go#L87-L162), except for:
- `tenant_id` is not supported by [current Synapse go SDK](https://github.com/Azure/azure-sdk-for-go/blob/af40ce6ee771cac2772597de756039d2d96a13de/services/preview/synapse/mgmt/2019-06-01-preview/synapse/models.go#L13643-L13658).
- `azure_devops_repo` instead of `vsts_configuration`, since the [official documentation](https://docs.microsoft.com/en-us/azure/synapse-analytics/cicd/source-control#connect-with-azure-devops-git) is using the term Azure DevOps.
- `github_repo` instead of `github_configuration`

### CODE REVIEW

**Comment by allantargino** [comment]
Hi @tombuildsstuff @katbyte, I would appreciate your review 😄

**Review by katbyte — CHANGES_REQUESTED** [review]
Hey @allantargino - just ran the tests for this but i am getting this error:
```
------- Stdout: -------
=== RUN   TestAccSynapseWorkspace_azdo
=== PAUSE TestAccSynapseWorkspace_azdo
=== CONT  TestAccSynapseWorkspace_azdo
    testing.go:684: Step 0 error: errors during apply:
        
        Error: Error retrieving ACLs for Root path in File System "acctest-210128063831382385" in Storage Account "acctestacc7j0nw": datalakestore.Client#GetProperties: Failure responding to request: StatusCode=403 -- Original Error: autorest/azure: error response cannot be parsed: "" error: EOF
        
          on /opt/teamcity-agent/temp/buildTmp/tf-test098443127/main.tf line 21:
          (source code not available)
        
        
    testing.go:745: Error destroying resource! WARNING: Dangling resources
        may exist. The full state and error is shown below.
        
        Error: errors during refresh: Error retrieving ACLs for Root path in File System "acctest-210128063831382385" in Storage Account "acctestacc7j0nw": datalakestore.Client#GetProperties: Failure responding to request: StatusCode=403 -- Original Error: autorest/azure: error response cannot be parsed: "" error: EOF
        
        State: <nil>
--- FAIL: TestAccSynapseWorkspace_azdo (51.03s)
FAIL
```

do you get this error?

File: azurerm/internal/services/synapse/synapse_workspace_resource.go [hunk 1]
```diff
@@ -146,6 +151,78 @@ func resourceSynapseWorkspace() *schema.Resource {
...
+						},
+						"branch_name": {
+							Type:         schema.TypeString,
+							Required:     true,
+							ValidateFunc: validation.StringIsNotEmpty,
+						},
+						"git_url": {
+							Type:         schema.TypeString,
+							Optional:     true,
+							ValidateFunc: validation.StringIsNotEmpty,
```
[the remark this answers is not in the document]
Comment (allantargino) [reply azurerm/internal/services/synapse/synapse_workspace_resource.go]: Ok! Thanks for the suggestion.

**Comment by allantargino** [comment]
Hi @katbyte,
I am not getting this error, I just ran the tests again:
image
image

**Comment by katbyte** [comment]
Looks like something changed on the 26th and all our tests started to fail for synapse with that error, will have to look into whats going on with our subscription

**Comment by njuCZ** [comment]
@katbyte recently I am doing another task about Synapse and also found this issue. It seems it's casued by retrieving ACL rest api. 
the response status code is 403, which shows the permission is not enough. After I assign the `Storage Blob Data Contributor` role and test again, it report 400 this time. It seems it's by design when the hns feature is disabled.
I have submit a PR to fix this problem: https://github.com/terraform-providers/terraform-provider-azurerm/pull/10470

File: azurerm/internal/services/synapse/synapse_workspace_resource.go [hunk 2]
```diff
@@ -146,6 +151,78 @@ func resourceSynapseWorkspace() *schema.Resource {
 				ValidateFunc: validate.ManagedResourceGroupName(),
 			},
 
+			"azure_devops_configuration": {
```
Comment (katbyte) [inline azurerm/internal/services/synapse/synapse_workspace_resource.go:154]: I think we can drop configuration
```suggestion
			"azure_devops_source": {
```

or
```suggestion
			"azure_devops_repo": {
```

File: azurerm/internal/services/synapse/synapse_workspace_resource.go [hunk 3]
```diff
@@ -146,6 +151,78 @@ func resourceSynapseWorkspace() *schema.Resource {
...
+						"root_folder": {
+							Type:         schema.TypeString,
+							Required:     true,
+							ValidateFunc: validate.RepoRootFolder(),
+						},
+					},
+				},
+			},
+
+			"github_configuration": {
```
Comment (katbyte) [inline azurerm/internal/services/synapse/synapse_workspace_resource.go:190]: and here

**Review by katbyte — CHANGES_REQUESTED** [review]
@allantargino - tests are fixed now, just one comment about the schema and this should be good to merge

File: azurerm/internal/services/synapse/synapse_workspace_resource.go (hunk 2 above)
Comment (allantargino) [reply azurerm/internal/services/synapse/synapse_workspace_resource.go:154]: `*_repo` sounds great!

**Comment by allantargino** [comment]
Thanks @katbyte, I applied the suggested changes.

**Review by katbyte — APPROVED** [review]
Thanks @allantargino - this LGTM now 👍

**Pull request merged**